### PR TITLE
Handling OpenSearchRejectExecuteException Exception

### LINKED ISSUE #627: [BUG] Replication stops with NotSerializableExceptionWrapper exception
**What is the bug?**

Replication stops with org.opensearch.common.io.stream.NotSerializableExceptionWrapper exception

```
{
  "status" : "PAUSED",
  "reason" : """AutoPaused: [[xx-rr][18] - org.opensearch.common.io.stream.NotSerializableExceptionWrapper - "replication_exception: failed to replay changes"], [[xx-rr][62] - org.opensearch.common.io.stream.NotSerializableExceptionWrapper - "replication_exception: failed to replay changes"], """,
  "leader_alias" : "xx",
  "leader_index" : "xx",
  "follower_index" : "xxrr"
}
```

And when I try to resume it ,

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "resource_not_found_exception",
        "reason" : "Retention lease doesn't exist. Replication can't be resumed for xxx-rr"
      }
    ],
    "type" : "resource_not_found_exception",
    "reason" : "Retention lease doesn't exist. Replication can't be resumed for xxx-rr"
  },
  "status" : 404
}
```

And it also does auto pause

```
{
  "status" : "PAUSED",
  "reason" : """AutoPaused: [[xx-rr-1][23] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@72fa7b3a on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10004, completed tasks = 4620]]"], [[xx-rr-1][86] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@45d9c393 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10011, completed tasks = 4620]]"], [[xx-rr-1][27] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@121fc524 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][58] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@73f98d3e on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10009, completed tasks = 4620]]"], [[xx-rr-1][15] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@5d4e3761 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][14] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@312f038e on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][46] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@ed8c638 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10001, completed tasks = 4620]]"], [[xx-rr-1][76] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4c9f9430 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10009, completed tasks = 4620]]"], [[xx-rr-1][82] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@470736e2 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10017, completed tasks = 4620]]"], [[xx-1][7] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@74f7db9d on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10002, completed tasks = 4620]]"], [[xx-rr-1][38] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@65cc9307 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][11] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@1924a8ae on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10004, completed tasks = 4620]]"], [[xx-rr-1][75] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@13110bfa on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[sxx-rr-1][72] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@ecacfe0 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10012, completed tasks = 4620]]"], [[xx-rr-1][61] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4f5cec55 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][3] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4a7e6d26 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][66] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@5777b9a9 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10002, completed tasks = 4620]]"], [[xx-rr-1][33] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@7765cd1d on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10011, completed tasks = 4620]]"], [[xx-rr-1][32] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4d7cb3b5 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], """,
  "leader_alias" : "xx-prod",
  "leader_index" : "xx",
  "follower_index" : "xx-1"
}
```

**What is your host/environment?**
 - OS: AWS Opensearch
 - Version 1.3
 - Plugins

### PR DESCRIPTION
**Pull request by sricharanvuppu**
### Description
This PR handles OpenSearchRejectExecuteException in suspendExecuteWithRetries
This PR also introduces new `plugins.replication.follower.concurrent_writers_per_shard` setting, this new setting gives user more control over concurrent writes. It default is 2 same as `concurrent_readers_per_shard` and can be changed to minimum value : 1  
### Issues Resolved
#627 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationTask.kt [hunk 1]
```diff
@@ -214,7 +214,7 @@ class ShardReplicationTask(id: Long, type: String, action: String, description:
         // Since this setting is not dynamic, setting update would only reflect after pause-resume or on a new replication job.
         val rateLimiter = Semaphore(replicationSettings.readersPerShard)
         val sequencer = TranslogSequencer(scope, replicationMetadata, followerShardId, leaderAlias, leaderShardId.indexName,
-                                          TaskId(clusterService.nodeName, id), client, indexShard.localCheckpoint, followerClusterStats)
+                                          TaskId(clusterService.nodeName, id), client, indexShard.localCheckpoint, followerClusterStats, replicationSettings.readersPerShard)
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationTask.kt:217]: Let's have one more setting for writerpershard with default value as `2` to initialize this value

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt [hunk 2]
```diff
@@ -73,26 +88,52 @@ class TranslogSequencer(scope: CoroutineScope, private val replicationMetadata:
                     var relativeStartNanos  = System.nanoTime()
                     val retryOnExceptions = ArrayList<Class<*>>()
                     retryOnExceptions.add(MappingNotAvailableException::class.java)
+                    var tooManyRequestsAcknowledgement = true
+                    try {
+                        while (tooManyRequestsAcknowledgement) {
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:93]: For the current changes, let's retry after the backoff and handle any errors within the translog sequencer.

File: src/main/kotlin/org/opensearch/replication/util/Extensions.kt [hunk 3]
```diff
@@ -122,19 +125,29 @@ suspend fun <Req: ActionRequest, Resp: ActionResponse> Client.suspendExecuteWith
...
+                retryException = e;
             } else {
                 throw e
             }
+        } catch (e: OpenSearchRejectedExecutionException) {
+            if(index < numberOfRetries-2) {
+                retryException = e;
+            }
+            else {
+                throw ReplicationException("OpensearchRejectExecutionException being thrown as ReplicationException", RestStatus.TOO_MANY_REQUESTS,e as Throwable)
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/util/Extensions.kt:137]: Could we construct the message from the exception?

File: src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationTask.kt (hunk 1 above)
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/task/shard/ShardReplicationTask.kt:217]: Ack

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt (hunk 2 above)
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:93]: Ack

File: src/main/kotlin/org/opensearch/replication/util/Extensions.kt (hunk 3 above)
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/util/Extensions.kt:137]: will Modify it

Comment (ankitkala) [reply src/main/kotlin/org/opensearch/replication/util/Extensions.kt:137]: Yep `OpensearchRejectExecutionException being thrown as ReplicationException` is not very helpful. Let's use the original exception.

File: src/main/kotlin/org/opensearch/replication/util/Extensions.kt [hunk 4]
```diff
@@ -122,19 +125,29 @@ suspend fun <Req: ActionRequest, Resp: ActionResponse> Client.suspendExecuteWith
...
+                retryException = e;
             } else {
                 throw e
             }
+        } catch (e: OpenSearchRejectedExecutionException) {
+            if(index < numberOfRetries-2) {
+                retryException = e;
+            }
+            else {
+                throw ReplicationException(e.message.toString(), RestStatus.TOO_MANY_REQUESTS,e as Throwable)
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/util/Extensions.kt:137]: why not use `Throwable` directly for constructing exception instead on `toString()`?

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt [hunk 5]
```diff
@@ -73,26 +90,55 @@ class TranslogSequencer(scope: CoroutineScope, private val replicationMetadata:
...
-                        replayResponse.shardInfo.failures.forEachIndexed { i, failure ->
-                            log.error("Failed replaying changes. Failure:$i:$failure}")
+                                val tookInNanos = System.nanoTime() - relativeStartNanos
+                                followerClusterStats.stats[followerShardId]!!.totalWriteTime.addAndGet(
+                                    TimeUnit.NANOSECONDS.toMillis(tookInNanos)
+                                )
+                                followerClusterStats.stats[followerShardId]!!.opsWritten.addAndGet(
+                                    replayRequest.changes.size.toLong()
+                                )
+                            } catch (e: OpenSearchException) {
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:125]: Can we extract the following code from extensions.kt below and reuse here?

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt [hunk 6]
```diff
@@ -50,20 +57,30 @@ class TranslogSequencer(scope: CoroutineScope, private val replicationMetadata:
                         private val followerShardId: ShardId,
                         private val leaderAlias: String, private val leaderIndexName: String,
                         private val parentTaskId: TaskId, private val client: Client, initialSeqNo: Long,
-                        private val followerClusterStats: FollowerClusterStats) {
-
+                        private val followerClusterStats: FollowerClusterStats, writersPerShard : Int) {
+    //Start backOff for exceptions with a second
+    private val initialBackoffMillis = 1000L
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:62]: Seems to be redundant based on the logic from Extensions.kt, Could we re-use the method here?

File: src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt [hunk 7]
```diff
@@ -347,11 +349,12 @@ internal class ReplicationPlugin : Plugin(), ActionPlugin, PersistentTaskPlugin,
     override fun getSettings(): List<Setting<*>> {
         return listOf(REPLICATED_INDEX_SETTING, REPLICATION_FOLLOWER_OPS_BATCH_SIZE, REPLICATION_LEADER_THREADPOOL_SIZE,
                 REPLICATION_LEADER_THREADPOOL_QUEUE_SIZE, REPLICATION_FOLLOWER_CONCURRENT_READERS_PER_SHARD,
-                REPLICATION_FOLLOWER_RECOVERY_CHUNK_SIZE, REPLICATION_FOLLOWER_RECOVERY_PARALLEL_CHUNKS,
-                REPLICATION_PARALLEL_READ_POLL_INTERVAL, REPLICATION_AUTOFOLLOW_REMOTE_INDICES_POLL_INTERVAL,
-                REPLICATION_AUTOFOLLOW_REMOTE_INDICES_RETRY_POLL_INTERVAL, REPLICATION_METADATA_SYNC_INTERVAL,
-                REPLICATION_RETENTION_LEASE_MAX_FAILURE_DURATION, REPLICATION_INDEX_TRANSLOG_PRUNING_ENABLED_SETTING,
-                REPLICATION_INDEX_TRANSLOG_RETENTION_SIZE, REPLICATION_FOLLOWER_BLOCK_START, REPLICATION_AUTOFOLLOW_CONCURRENT_REPLICATION_JOBS_TRIGGER_SIZE)
+                REPLICATION_FOLLOWER_CONCURRENT_WRITERS_PER_SHARD, REPLICATION_FOLLOWER_RECOVERY_CHUNK_SIZE,
```
[the remark this answers is not in the document]
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/ReplicationPlugin.kt:352]: Ack

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt (hunk 5 above)
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:125]: missed java.class exceptions will add them as mentioned in extenions.kt

**Review by saikaranam-amazon — APPROVED** [review]

**Review by ankitkala — APPROVED** [review]

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt [hunk 8]
```diff
@@ -59,11 +64,14 @@ class TranslogSequencer(scope: CoroutineScope, private val replicationMetadata:
     val followerIndexService = indicesService.indexServiceSafe(followerShardId.index)
     val indexShard = followerIndexService.getShard(followerShardId.id)
 
-    private val sequencer = scope.actor<Unit>(capacity = Channel.UNLIMITED) {
+    private val sequencer = scope.actor<Unit>(capacity = 0) {
+
         // Exceptions thrown here will mark the channel as failed and the next attempt to send to the channel will
         // raise the same exception.  See [SendChannel.close] method for details.
+        val rateLimiter = Semaphore(writersPerShard)
```
[the remark this answers is not in the document]
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:71]: Ack

**Pull request merged**